### ISSUE TITLE
missing units

### ISSUE THREAD

**Issue by aducom** [Alexey-T/ATSynEdit_Ex#3]
Probably the docs on lazarus wiki are out of date, but from the wiki:
Unit atsynedit_form_complete has function to show completion-listbox for editor. Listbox is not modal. It keeps editor caret blinking. While listbox shown, editor don't get keyboard input, but listbox passes typed chars/Backspace/Left/Right to editor. Listbox replaces text on Enter/Tab keys, hides on Esc or clicking outside.

I'm trying to implement a code-completion, but this unit is not there. I also tried an implementation with econtrol (I'm not sure if it should work, as I know I need a license for my use, but first I want to see it in action). 

The mentioned unit is not on the download here, I could find an old version that does, but I got totally stuck there (and it's older than your version).

Is there a sample containing code completion somewhere?

**Comment by Alexey-T** [comment]
That unit is not in ATsynedit, it’s in the Lazarus package atsynedit_cmp. This package must be installed. Did you get it?

**Comment by Alexey-T** [comment]
I corrected this wiki part now.

**Comment by aducom** [comment]
It's not there:
![image](https://user-images.githubusercontent.com/5610893/132709766-7a8aec25-c4b2-463c-b8a2-3e5e7c2fd898.png)

**Comment by aducom** [comment]
But perhaps I'm doing something wrong also:
`  { TFEditor }
  TATCompletionPropEvent = procedure (Sender: TObject;
    out AText: string; out ACharsLeft, ACharsRight: integer) of object;

  TFEditor = class(TForm)
...
 procedure DoEditorCompletionListbox(AEd: TATSynEdit; AOnGetProp: TATCompletionPropEvent);
         procedure FinderUpdateEditor(AUpdateText: boolean);

`

In the implementation 
`procedure TFEditor.DoEditorCompletionListbox(AEd: TATSynEdit; AOnGetProp: TATCompletionPropEvent);
var a : string;
begin
   //AText:= 'func|Func1|(param1, param2)'#9'Func desc' + #13 + 'var|Var1' + #13 + 'var|Var2';
end;
`

Compiles to an error about AText. And the event is not triggered. But I can't find a setting to enable this callback and could not find the root method... (abstract nor virtual)

**Comment by Alexey-T** [comment]
https://wiki.freepascal.org/ATSynEdit#General_API

The wiki part was outdated and wrong. I fixed - function name and signature were wrong.

**Comment by aducom** [comment]
Thank you, I'll dive into it.

**Comment by aducom** [comment]
Ok, I added the function types, created 
`  TATCompletionPropEvent = procedure (Sender: TObject;
    AContent: TStringList; out ACharsLeft, ACharsRight: integer) of object;
  TATCompletionResultEvent = procedure (Sender: TObject;
    const ASnippetId: string; ASnippetIndex: integer) of object;

  TFEditor = class(TForm)      `

As these are types, I added two variables
`     CompletePropEvent : TATCompletionPropEvent;
     CompleteResultEvent : TATCompletionResultEvent;
`
Actually I think I have to use a property of one of the objects, but don't know which.

Added declaration:
`
    procedure EditorShowCompletionListbox(AEd: TATSynEdit;  AOnGetProp: TATCompletionPropEvent;  AOnResult: TATCompletionResultEvent = nil;
                                                            AOnChoose: TATCompletionResultEvent = nil; const ASnippetId: string = '';  ASelectedIndex: integer = 0; AAllowCarets: boolean = false);

    procedure CompletionPropEvent(Sender: TObject; AContent: TStringList; out ACharsLeft, ACharsRight: integer);
    procedure CompletionResultEvent(Sender: TObject; const ASnippetId: string; ASnippetIndex: integer);
                                  `
`
I added the following (needed a simple statement so I can set a breakpoint)
`procedure TFEditor.EditorShowCompletionListbox(AEd: TATSynEdit;  AOnGetProp: TATCompletionPropEvent;  AOnResult: TATCompletionResultEvent = nil;
                                                         AOnChoose: TATCompletionResultEvent = nil; const ASnippetId: string = '';  ASelectedIndex: integer = 0; AAllowCarets: boolean = false);
var a : integer;
begin
   a:= 1;
end;

procedure TFEditor.CompletionPropEvent(Sender: TObject; AContent: TStringList; out ACharsLeft, ACharsRight: integer);
var a: integer;
begin
   a := 1;
end;

procedure TFEditor.CompletionResultEvent(Sender: TObject; const ASnippetId: string; ASnippetIndex: integer);
var a : integer;
begin
   a:=1;
end;
`
It compiles, but does not get triggered. So I need to 'tell' the editor that these methods are in place, so I probably have to enable/disable something?

**Comment by Alexey-T** [comment]
a) I don't understand why you ADD my declarations to your code??? just add "uses UnitNameHere" to use my unit.
b) better give the compilable project in lazarus to me.

**Comment by aducom** [comment]
Well, sorry for that, I did not understand the docs correctly then. Been strugling all day so I'm not seeing things clearly any more. But yes, all is in your code. But setting breakpoint proves that it is not triggered. My source is based upon the sample, but I don't mind share it to you. But perhaps you can consider adding it to one of your samples?

**Comment by Alexey-T** [comment]
I will make the auto completion demo, pls wait some days.

**Comment by aducom** [comment]
Would be very helpful. Thnx.Op 9 sep. 2021 19:48 schreef "Alexey T." ***@***.***>:
I will make the auto completion demo, pls wait some days.

**Comment by Alexey-T** [comment]
I added simple demo! to ATSynedit_Cmp package.
Try it--
https://github.com/Alexey-T/ATSynEdit_Cmp/tree/master/app/demo_simple

**Comment by aducom** [comment]
Looks nice. To get your sample running I had to create a new project as I could not load it in lazarus 2.0.12. Then I got some missing constants and I installed your latest version. Added the events and now it works nice!

Other remark is that the DoCommand has changed. So in my original application I now have:

  Edit.DoCommand(cCommand_ScrollToCaretTop);

Which is generating an error, I need an extra parameter but which? I can find a handful, cInvokeInternal?
Should I send you my version? (actually exactly the same, but loads in Lazarus).
